Python/import.c

```diff
@@ -952,134 +973,125 @@ static int
 _extensions_cache_set(PyObject *filename, PyObject *name, PyModuleDef *def)
 {
     int res = -1;
-    PyThreadState *oldts = NULL;
     extensions_lock_acquire();
 
-    /* Swap to the main interpreter, if necessary.  This matters if
-       the dict hasn't been created yet or if the item isn't in the
-       dict yet.  In both cases we must ensure the relevant objects
-       are created using the main interpreter. */
-    PyThreadState *main_tstate = &EXTENSIONS.main_tstate;
-    PyInterpreterState *interp = _PyInterpreterState_GET();
-    if (!_Py_IsMainInterpreter(interp)) {
-        _PyThreadState_BindDetached(main_tstate);
-        oldts = _PyThreadState_Swap(interp->runtime, main_tstate);
-        assert(!_Py_IsMainInterpreter(oldts->interp));
-
-        /* Make sure the name and filename objects are owned
-           by the main interpreter. */
-        name = PyUnicode_InternFromString(PyUnicode_AsUTF8(name));
-        assert(name != NULL);
-        filename = PyUnicode_InternFromString(PyUnicode_AsUTF8(filename));
-        assert(filename != NULL);
+    if (EXTENSIONS.hashtable == NULL) {
+        EXTENSIONS.hashtable = _Py_hashtable_new_full(
```

Comment:
Grr, I think GH ate my comment. (Apologies if this is a repeat and I just didn't see it.)

Creating a two-level hashtable gives me the heebie-jeebies. _By far_ the most common case is for the filename to be unique, so all this effort is really going to waste. (I say this as someone who maintains infrastructure that routinely uses the same "filename" for multiple extension modules; I know it's really uncommon.) Can't we just use the fact that filenames won't be directories, and names won't contain path separators, and make a single hashtable with the key being `filename + '/' + name` instead? (I know module names _could_ contain / but since this name is also used for the PyInit function, that wouldn't work for extension modules anyway.) You can still reliably split them for debugging, but none of Python actually cares about recovering filename/modulename from the cache key.

Reply:
> Creating a two-level hashtable gives me the heebie-jeebies.

Yeah, me too.  At the time the only alternatives I could think of felt *more* complicated.  I'll gladly take a less icky approach. 😄

> make a single hashtable with the key being filename + '/' + name instead?

I had considered that after I'd done the nested hashtable but thought there was some extra complexity to the hashing and didn't get back to it.  I think this should work though and will give it a try.

Reply:
done

Reply:
Done!  It's definitely cleaner. 😄